src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,284 @@
+    Printf.@printf(io, "%.2f", m.lp)
+    println(io)
+    show(io, m.values)
+end
+
+function Base.show(io::IO, m::ModeResult)
+    show(io, m.values.array)
+end
+
+function StatsBase.coeftable(m::ModeResult)
+    # Get columns for coeftable.
+    terms = StatsBase.coefnames(m)
+    estimates = m.values.array[:,1]
+    stderrors = StatsBase.stderror(m)
+    tstats = estimates ./ stderrors
+
+    StatsBase.CoefTable([estimates, stderrors, tstats], ["estimate", "stderror", "tstat"], terms)
+end
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
+    info = inv(hessian_function(m.f, m.values.array[:, 1]))
+    return NamedArrays.NamedArray(info, (varnames, varnames))
+end
+
+StatsBase.coef(m::ModeResult) = m.values
+StatsBase.coefnames(m::ModeResult) = names(m.values)[1]
+StatsBase.params(m::ModeResult) = StatsBase.coefnames(m)
+StatsBase.vcov(m::ModeResult) = StatsBase.informationmatrix(m)
+StatsBase.loglikelihood(m::ModeResult) = m.lp
+
+####################
+# Optim.jl methods #
+####################
+
+"""
+    Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+
+Compute a maximum likelihood estimate of the `model`.
+
+# Examples
+
+```julia-repl
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+mle = optimize(model, MLE())
+
+# Use a different optimizer
+mle = optimize(model, MLE(), NelderMead())
+```
+"""
+function Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+    ctx = OptimizationContext(DynamicPPL.LikelihoodContext())
+    return optimize(model, OptimLogDensity(model, ctx), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+
+Compute a maximum a posterior estimate of the `model`.
+
+# Examples
+
+```julia-repl
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+map_est = optimize(model, MAP())
+
+# Use a different optimizer
+map_est = optimize(model, MAP(), NelderMead())
+```
+"""
+function Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+    ctx = OptimizationContext(DynamicPPL.DefaultContext())
+    return optimize(model, OptimLogDensity(model, ctx), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+
+Estimate a mode, i.e., compute a MLE or MAP estimate.
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+    # Do some initialization.
+    spl = DynamicPPL.SampleFromPrior()
+    init_vals = f.vi[spl]
+
+    # Optimize!
+    M = Optim.optimize(f, init_vals, optimizer, args...; kwargs...)
+
+    # Get the VarInfo at the MLE/MAP point, and run the model to ensure 
+    # correct dimensionality.
+    f.vi[spl] = M.minimizer
+    invlink!(f.vi, spl)
+    vals = f.vi[spl]
+    link!(f.vi, spl)
```

Comment:
This means the hessian will not be wrt to the constrained parameters?

Reply:
Yup, you're right. This is what I get for relying on bad behavior elsewhere in the code.